### PR TITLE
CODETOOLS-7903471: JMH: Add test for virtual threads executor

### PR DESCRIPTION
**Pull request by kuksenko**
Add test for virtual thread executor

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [CODETOOLS-7903471](https://bugs.openjdk.org/browse/CODETOOLS-7903471): JMH: Add test for virtual threads executor

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmh.git pull/103/head:pull/103` \
`$ git checkout pull/103`

Update a local copy of the PR: \
`$ git checkout pull/103` \
`$ git pull https://git.openjdk.org/jmh.git pull/103/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 103`

View PR using the GUI difftool: \
`$ git pr show -t 103`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmh/pull/103.diff">https://git.openjdk.org/jmh/pull/103.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jmh/pull/103#issuecomment-1548978398)

### CODE REVIEW

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.it.Fixtures;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
```
Comment (shipilev) [inline jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java:46]: Please check if we need all these imports?

**Review by shipilev — CHANGES_REQUESTED** [review]
Suggestions:

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Tests if harness executes setup, run, and tearDown in the same workers.
```
[the remark this answers is not in the document]
Comment (kuksenko) [reply jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java:56]: Yeah. When code speaks more than comments, comments become invisible. :)

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    private final Set<Boolean> setupIterationThread = Collections.synchronizedSet(new HashSet<>());
+    private final Set<Boolean> setupInvocationThread = Collections.synchronizedSet(new HashSet<>());
+    private final Set<Boolean> teardownRunThread = Collections.synchronizedSet(new HashSet<>());
+    private final Set<Boolean> teardownIterationThread = Collections.synchronizedSet(new HashSet<>());
+    private final Set<Boolean> teardownInvocationThread = Collections.synchronizedSet(new HashSet<>());
+    private final Set<Boolean> testInvocationThread = Collections.synchronizedSet(new HashSet<>());
+
+    @Setup(Level.Trial)
+    public void setupRun() {
+        setupRunThread.add(VirtualAPI.isVirtual(Thread.currentThread()));
```
[the remark this answers is not in the document]
Comment (kuksenko) [reply jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java:74]: There are other tests that check a single expected thread.

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java (hunk 1 above)
Comment (kuksenko) [reply jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java:46]: yes

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kuksenko) [reply jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java]: Ok.
Also, that means - we don't need this new test class. It's better just to add isVirtual check into existing ones.

[the remark this answers is not in the document]
Comment (kuksenko) [reply jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchExecutorTest.java]: and now?

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchmarkBenchSameThreadTest.java [hunk 4]
```diff
@@ -96,12 +120,40 @@ public void tearDownInvocation() {
     @TearDown(Level.Trial)
     public void teardownZZZ() { // should perform last
         Assert.assertFalse("Test sanity", testInvocationThread.isEmpty());
-        Assert.assertTrue("test <: setupRun", testInvocationThread.containsAll(setupRunThread));
-        Assert.assertTrue("test <: setupIteration", testInvocationThread.containsAll(setupIterationThread));
-        Assert.assertTrue("test <: setupInvocation", testInvocationThread.containsAll(setupInvocationThread));
-        Assert.assertTrue("test <: teardownRun", testInvocationThread.containsAll(teardownRunThread));
-        Assert.assertTrue("test <: teardownIteration", testInvocationThread.containsAll(teardownIterationThread));
-        Assert.assertTrue("test <: teardownInvocation", testInvocationThread.containsAll(teardownInvocationThread));
+        if(benchmarkExecutorType != ExecutorType.CACHED_TPE) { // CachedThreadPool doesn't guarantee same thread rule
```
Comment (shipilev) [inline jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchmarkBenchSameThreadTest.java:123]:
```suggestion
        if (benchmarkExecutorType != ExecutorType.CACHED_TPE) { // CachedThreadPool doesn't guarantee same thread rule
```

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchmarkBenchSameThreadTest.java [hunk 5]
```diff
@@ -96,12 +120,40 @@ public void tearDownInvocation() {
...
+            Assert.assertTrue("test <: setupRun", testInvocationThread.containsAll(setupRunThread));
+            Assert.assertTrue("test <: setupIteration", testInvocationThread.containsAll(setupIterationThread));
+            Assert.assertTrue("test <: setupInvocation", testInvocationThread.containsAll(setupInvocationThread));
+            Assert.assertTrue("test <: teardownRun", testInvocationThread.containsAll(teardownRunThread));
+            Assert.assertTrue("test <: teardownIteration", testInvocationThread.containsAll(teardownIterationThread));
+            Assert.assertTrue("test <: teardownInvocation", testInvocationThread.containsAll(teardownInvocationThread));
+        }
+        if(benchmarkExecutorType == ExecutorType.VIRTUAL_TPE) {
```
Comment (shipilev) [inline jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchmarkBenchSameThreadTest.java:131]:
```suggestion
        if (benchmarkExecutorType == ExecutorType.VIRTUAL_TPE) {
```

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchmarkBenchSameThreadTest.java [hunk 6]
```diff
@@ -96,12 +120,40 @@ public void tearDownInvocation() {
...
+        } else {
+            Assert.assertTrue("setupRun thread kind", setupRunThread.stream().noneMatch(VirtualAPI::isVirtual));
+            Assert.assertTrue("setupIteration thread kind", setupIterationThread.stream().noneMatch(VirtualAPI::isVirtual));
+            Assert.assertTrue("setupInvocation thread kind", setupInvocationThread.stream().noneMatch(VirtualAPI::isVirtual));
+            Assert.assertTrue("teardownRun thread kind", teardownRunThread.stream().noneMatch(VirtualAPI::isVirtual));
+            Assert.assertTrue("teardownIteration thread kind", teardownIterationThread.stream().noneMatch(VirtualAPI::isVirtual));
+            Assert.assertTrue("teardownInvocation thread kind", teardownInvocationThread.stream().noneMatch(VirtualAPI::isVirtual));
+            Assert.assertTrue("testInvocation thread kind", testInvocationThread.stream().noneMatch(VirtualAPI::isVirtual));
+        }
+        if(benchmarkExecutorType == ExecutorType.FJP) {
```
Comment (shipilev) [inline jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchmarkBenchSameThreadTest.java:148]:
```suggestion
        if (benchmarkExecutorType == ExecutorType.FJP) {
```

File: jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchmarkBenchSameThreadTest.java [hunk 7]
```diff
@@ -116,14 +168,201 @@ public void test() {
...
+        }
+    }
+
+    @Test
+    public void invokeAPI_custom() throws RunnerException {
+        for (int c = 0; c < Fixtures.repetitionCount(); c++) {
+            Options opt = new OptionsBuilder()
+                    .include(Fixtures.getTestMask(this.getClass()))
+                    .jvmArgsAppend("-Djmh.executor=CUSTOM")
+                    .jvmArgsAppend("-Djmh.executor.class="+CustomExecutor.class.getName())
```
Comment (shipilev) [inline jmh-core-it/src/test/java/org/openjdk/jmh/it/threads/BenchmarkBenchSameThreadTest.java:226]:
```suggestion
                    .jvmArgsAppend("-Djmh.executor.class=" + CustomExecutor.class.getName())
```

**Review by shipilev — APPROVED** [review]
Looks fine with minor nits:

**Review by shipilev — CHANGES_REQUESTED** [review]
Actually, no, wait. See GHA failures, there are plenty of them.

**Comment by kuksenko** [comment]
> Actually, no, wait. See GHA failures, there are plenty of them.

I am working on it. It's intermittent failures, which I can't repeat locally.

**Comment by kuksenko** [comment]
@shipilev  We need to discuss this: https://bugs.openjdk.org/browse/CODETOOLS-7903476
Either a large threading redesign should be done or simply remove CACHED_TPE, FJP_COMMON, and CUSTOM.

**Comment by kuksenko** [comment]
Checking how works. Also fixing CODETOOLS-7903476

**Comment by kuksenko** [comment]
Q.E.D.
Benchmarks execution requires many synchronization points between benchmark threads (Not even counting the new synch point I added to solve the "orphaned" worker data issue). Benchmark executor should provide N (a.k.a. -t N) distinct threads; otherwise, a deadlock will happen. 
All existing executors provide this except FJP_COMMON.  
Now I see two options:
- Completely remove FJP_COMMON 
- Preserve FJP_COMMON for the user's risk, but don't create tests covering FJP_COMMON.

**Comment by shipilev** [comment]
I think this is almost completely subsumed by #107. We can repurpose this to check for `isVirtual`, if you want. `Fixtures` helper method can sense the testing profile we are running with, `executor-virtual` would be the thing to check for. I don't see significant value in checking for `isVirtual`, though. Your call.

**Comment by shipilev** [comment]
> Agree. I think proper fix and new tests for stable threads are more comprehensive. I am ready to close this PR without integration.

OK, feel free to close this PR and the JIRA issue for it.

**Comment by kuksenko** [comment]
The test is no more relevant. https://github.com/openjdk/jmh/pull/107 gives better coverage.

**Pull request closed without merging**